### PR TITLE
Vocabularies

### PR DESCRIPTION
**Pull request by cgalarza**
This is a start /vocabularies endpoint. Error reporting, etc needs to be improved. The model for Vocabulary is also intentionally missing a has_many terms relationship.

OPTIONS /api/vocabularies has not been implemented.

### CODE REVIEW

**Comment by cgalarza** [comment]
@elohanlon Hey, this is ready for you to review!

File: app/controllers/v1/vocabularies_controller.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      vocabularies = Vocabulary.all
+      render json: { vocabularies: vocabularies.map { |v| v.as_json(only: SHOW_FIELDS) } }.to_json
+    end
+
+    # GET /vocabularies/:string_key
+    def show
+      if (vocabulary = Vocabulary.find_by(string_key: params[:string_key]))
+        render json: vocabulary.to_json(only: SHOW_FIELDS), status: 200
+      else
+        render json: { error: { message: 'Not Found' } }.to_json, status: 404
```
Comment (elohanlon) [inline app/controllers/v1/vocabularies_controller.rb]: Based on the new JSON-API-like error format we talked about (for this, and for other controller error responses):
```
{
  errors: [
    {
      title: 'Not Found'
    }
  ]
}
```
See: https://jsonapi.org/format/#error-objects

File: app/controllers/v1/vocabularies_controller.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      end
+    end
+
+    # POST /vocabularies
+    def create
+      vocabulary = Vocabulary.new(create_params)
+      if vocabulary.save
+        render json: vocabulary.to_json(only: SHOW_FIELDS), status: 201
+      else
+        render json: { error: { messages: vocabulary.errors.full_messages.to_sentence } }, status: 400
```
Comment (elohanlon) [inline app/controllers/v1/vocabularies_controller.rb]: See above JSON API error comment.

File: app/controllers/v1/vocabularies_controller.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    # PATCH /vocabularies/:string_key
+    def update
+      vocabulary = Vocabulary.find_by(string_key: params[:string_key])
+
+      if vocabulary.nil?
+        render json: { error: { message: 'Not Found' } }, status: 404
+      elsif vocabulary.update(update_params)
+        render json: vocabulary.to_json(only: SHOW_FIELDS), status: 200
+      else
+        render json: { error: { messsage: vocabulary.errors.full_messages.to_sentence, code: 400 } }, status: 400
```
Comment (elohanlon) [inline app/controllers/v1/vocabularies_controller.rb]: See above JSON API error comment.

File: app/controllers/v1/vocabularies_controller.rb [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        render json: { error: { messages: vocabulary.errors.full_messages.to_sentence } }, status: 400
+      end
+    end
+
+    # PATCH /vocabularies/:string_key
+    def update
+      vocabulary = Vocabulary.find_by(string_key: params[:string_key])
+
+      if vocabulary.nil?
+        render json: { error: { message: 'Not Found' } }, status: 404
```
Comment (elohanlon) [inline app/controllers/v1/vocabularies_controller.rb]: See above JSON API error comment.

File: app/controllers/v1/vocabularies_controller.rb [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    # DELETE /vocabularies/:string_key
+    def destroy
+      vocabulary = Vocabulary.find_by(string_key: params[:string_key])
+
+      if vocabulary.nil?
+        render json: { error: { message: 'Not Found' } }, status: 404
+      elsif vocabulary.destroy
+        render json: '', status: :no_content
+      else
+        render json: { error: { message: 'Deleting was unsuccessful.' } }, status: 500
```
Comment (elohanlon) [inline app/controllers/v1/vocabularies_controller.rb]: See above JSON API error comment.

File: app/controllers/v1/vocabularies_controller.rb [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    end
+
+    # DELETE /vocabularies/:string_key
+    def destroy
+      vocabulary = Vocabulary.find_by(string_key: params[:string_key])
+
+      if vocabulary.nil?
+        render json: { error: { message: 'Not Found' } }, status: 404
+      elsif vocabulary.destroy
+        render json: '', status: :no_content
```
Comment (elohanlon) [inline app/controllers/v1/vocabularies_controller.rb]:
```suggestion
        `head :no_content`
```

☝️ I think you can do this instead.

**Review by elohanlon — APPROVED** [review]

**Pull request merged**